archive/android_bootldr.ksy

```diff
@@ -0,0 +1,36 @@
+meta:
+  id: android_bootldr
+  title: Qualcomm Snapdragon (MSM) bootloader.img format
+  file-extension: img
+  tags:
+    - archive
+    - android
+  license: CC0-1.0
+  endian: le
+doc-ref: https://android.googlesource.com/device/lge/hammerhead/+/7618a7/releasetools.py
+seq:
+  - id: magic
+    contents: BOOTLDR!
+  - id: images
+    -orig-id: num_images
+    type: u4
+  - id: ofs_start
+    -orig-id: start_offset
+    type: u4
+  - id: bootloader_size
+    -orig-id: bootldr_size
+    type: u4
+  - id: img_info
+    type: img_info
+    repeat: expr
+    repeat-expr: images
+types:
+  img_info:
+    -webide-representation: '{name}'
+    seq:
+      - id: name
+        type: strz
+        encoding: UTF-8
+        size: 64
+      - id: size
+        type: u4
```

Comment:
1. I'd rename it to `len_body`, since just calling it `size` does not necessarily say what size it is.

   ```suggestion
         - id: len_body
           type: u4
   ```

2. You could add an `instance` providing access to all byte arrays corresponding to each particular image:

   ```diff
      - id: img_infos
        type: img_info
        repeat: expr
        repeat-expr: num_images
   +instances:
   +  img_bodies_wrapper:
   +    pos: ofs_img_bodies
   +    size: len_img_bodies
   +    type: img_bodies_type
   +  img_bodies:
   +    value: img_bodies_wrapper.img_bodies
   +  is_valid:
   +    value: img_bodies_wrapper.is_eof
    types:
      img_info:
   @@ ... @@
   +  img_bodies_type:
   +    seq:
   +      - id: img_bodies
   +        size: _root.img_infos[_index].size
   +        repeat: expr
   +        repeat-expr: _root.num_images
   +      - size: 0
   +        if: is_eof < 0
   +    instances:
   +      is_eof:
   +        value: _io.eof
   ```
   
   _(uses previous suggestions on renaming the attribute `img_info` &rarr; <code>img_info**s**</code>, `'s/\<images\>/num_images/g'`, `'s/\<ofs_start\>/ofs_img_bodies/g'` and `'s/\<bootloader_size\>/len_img_bodies/g'`)._

   I consider the `is_valid` instance desirable because the Python parser linked in [`/doc-ref`](https://github.com/kaitai-io/kaitai_struct_formats/blob/d2500791ffe1b3d8b6af58552d3ddb275e2beb1b/archive/android_bootldr.ksy#L10) is really strict about `len_img_bodies`, originally `bootloader_size` ([`releasetools.py:162`](https://android.googlesource.com/device/lge/hammerhead/+/7618a7/releasetools.py#162)):

   ```python
     p = start_offset
     img_dict = {}
     for name, size in imgs:
       img_dict[trunc_to_null(name)] = p, size
       p += size
     assert p - start_offset == bootloader_size, "bootloader.img corrupted"
   ```

   To ensure that this KSY implementation is compatible with the reference parser, it should check the same thing - that's what the `is_valid` attribute is for. The check for `_io.eof` ensures that there are no additional bytes in the `len_img_bodies`-byte substream after the last image body.

Reply:
I am hesitant to use `_io.eof` because it complicates carving if there actually are additional bytes.